Safe module tutorial (#670)
safe-global/safe-docs: pages/advanced/smart-account-modules/smart-account-modules-tutorial.mdx

import { Callout } from 'nextra/components'

> # How to build an app with Safe and Safe Module

Reviewer: ```suggestion
# How to build an app with Safe and Safe Modules
```

should this be plural because it is about modules as a concept?

> In this tutorial, you will build an app that can:

Reviewer: ```suggestion
In this tutorial, you will:
```

the ending doesn't go well with the points. "build an app that can build a safe module"

- Build a Safe Module
- Enable an example module on a Safe (the [TokenTransferModule](https://github.com/5afe/safe-module-tutorial-contracts) module)
- Send a transaction via the TokenTransferModule (Send a transaction that transfers ERC20 token from the Safe account)

The full code for this tutorial is in the [Safe Module tutorial repository](https://github.com/5afe/safe-module-tutorial).

## Prerequisites

**Prerequisite knowledge:** You will need some basic experience with [React](https://react.dev/learn), [Next.js](https://nextjs.org/docs), [Solidity](https://docs.soliditylang.org/en/latest/) and [Hardhat](https://hardhat.org).

Before progressing with the tutorial, please make sure you have the following:

> - Downloaded and installed [Node.js](https://nodejs.org/en/download/package-manager) and [pnpm](https://pnpm.io/installation).

Reviewer: here it mentions pnpm but then throughout the tutorial `npm` is used

- Metamask installed in your browser and connected to the Sepolia network.
- Two test accounts in Metamask, the second with some Sepolia Eth for gas.

## Setup contracts repository

> This tutorial will provide a guide to create a Safe Module that allows users to withdraw ERC20 tokens from a Safe account, provided that the Safe owners have given their approval in the form of off-chain signatures. The primary use case is to enable beneficiaries to withdraw tokens themselves without requiring the Safe owners to execute transactions directly.

Reviewer: ```suggestion
This tutorial will guide creating a Safe Module that allows users to withdraw ERC20 tokens from a Safe account, provided that the Safe owners have given their approval in the form of off-chain signatures. The primary use case is to enable beneficiaries to withdraw tokens themselves without requiring the Safe owners to execute transactions directly.
```

### Use cases
- Simplifies token withdrawal by enabling beneficiaries to perform withdrawals themselves.
- Reduces the need for Safe owners to execute each transaction on-chain.

### Limitations
- Beneficiaries must be externally owned accounts (EOAs) and cannot be smart contracts.
- The nonce is sequential for simplicity
- The contract will work only for specific token and Safe address provided during deployment.

### Initialize a new repository

```bash
git init
```

```bash

> npm init

Reviewer 2: Could you add a line like "you can choose all default values". 

Also, it would be nice to have a line that says, "add npm-modules to .gitignore." Or you leave out the `git init` command.

```

### Install dependencies

Add overrides in `package.json` so that there are no peer dependency related issues.

```json
...

>   "overrides": {

Reviewer: I think you can include a more specific overide:
```
  "overrides": {
    "@safe-global/safe-contracts": {
      "ethers": "^6.13.5"
    }
  }
```

    "ethers": "^6.13.5"
  }
...
```

```bash
npm add hardhat @safe-global/safe-contracts @openzeppelin/contracts
```

### Initialize hardhat project

> Select create TypeScript project.

Reviewer: should be after calling the init command right?

Reviewer 2: yes.

Reviewer 2: The other options to choose should be namen, too.

(If it is possible to start with an empty hardhat.config, because it gets overwritten in the next step, this option should be chosen because it is easiest.

```bash
npx hardhat init
```

```bash
npx hardhat compile
```

### Update hardhat.config.ts

> When compiling Safe contracts with solidity 0.8.x the bytecode size exceeds the limit of 24KB. To overcome this, set `allowUnlimitedContractSize` to `true` in the hardhat config.

Reviewer: this means after the tutorial the contract will not be deployable to public networks, not sure if this is a good experience

Author: In practice, users will only deploy module code and not the Safe contracts.

Author: added the comment to use official safe addresses in production.

> Also, add `dependencyCompiler` to import `SafeProxyFactory` contract.

Reviewer 2: I had to run for the dependency compiler. Does this need to be added to the install command?
```npm install --save-dev hardhat-dependency-compiler```

Reviewer 2: I mean the npm install command at the very beginning.

- **`safeAddress`**: Stores the Safe contract address.
- **`tokenAddress`**: Stores the ERC20 token contract address.
- **`nonces`**: Tracks unique nonce to prevent replay attacks.

#### Step 4: Create the Constructor

Define a constructor to initialize the Safe and token contract addresses.

```solidity
constructor(address _tokenAddress, address _safeAddress) {
    tokenAddress = _tokenAddress;
    safeAddress = _safeAddress;
}
```

- Initializes `tokenAddress` and `safeAddress` with provided values during deployment. Thus, in this module the token and Safe addresses are fixed.

#### Step 5: Implement the `tokenTransfer` function

Add a function to handle token transfers from the Safe.

```solidity
function tokenTransfer(
    uint _amount,
    address _beneficiary,
    uint256 _deadline,
    bytes memory _signatures
) public {
    require(_deadline >= block.timestamp, "expired deadline");

    bytes32 signatureData = keccak256(
        abi.encode(
            PERMIT_TYPEHASH,
            _amount,
            msg.sender,
            nonces[msg.sender]++,
            _deadline
        )
    );

    bytes32 hash = keccak256(

>         abi.encodePacked("\x19\x01", getDomainSeparator(), signatureData)

Reviewer: `getDomainSeparator` is used here but only added in the next step. I'd define everything that's needed for signing first

            tokenAddress,
            0,
            data,
            Enum.Operation.Call
        ),
        "Could not execute token transfer"
    );
}
```

Explanation:
1. **Parameter Validation**:
   - Ensure the `_deadline` is valid.
   - Construct `signatureData` with the provided details and `PERMIT_TYPEHASH`.
2. **Hash Calculation**:
   - Compute the hash using the `EIP712` format to ensure signature consistency.
3. **Signature Verification**:
   - Call `checkSignatures` on the Safe to verify the signatures provided match the owners of the Safe.
4. **Transaction Execution**:
   - A module can use `execwhenTransactionFromModule` or `execTransactionFromModuleReturnData` function to execute transactions through a Safe account on which the module is enabled.
   - Encode the transfer call using `abi.encodeWithSignature`.
   - Use `execTransactionFromModule` to execute the token transfer via the Safe.
   - Ensure execution succeeds, otherwise revert.

#### Step 6: Implement the `getDomainSeparator` function

Add a helper function to compute the EIP-712 domain separator.

```solidity
function getDomainSeparator() private view returns (bytes32) {

>     uint256 chainId;

Reviewer: in 0.8.x block.chainid can be used

    assembly {
        chainId := chainid()
    }

    return keccak256(
        abi.encode(
            keccak256(
                "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
            ),
            keccak256(bytes("TokenWithdrawModule")),
            keccak256(bytes("1")),
            chainId,
            address(this)
        )
    );
}
```

Explanation:
- Computes the `EIP712Domain` separator for the current chain and contract.
- Ensures compatibility with the EIP-712 standard for off-chain signing.
- Using a Domain separator ensures that the signature is valid for specific contracts in context and the chain. Thus, preventing replay attacks.

#### Final contract code

> Here is the complete code for reference:

Reviewer 2: add "with comments" or so to indicate comments are added.

            _signatures
        );

        bytes memory data = abi.encodeWithSignature(
            "transfer(address,uint256)",
            _beneficiary,
            _amount
        );

        require(
            Safe(payable(safeAddress)).execTransactionFromModule(
                tokenAddress,
                0,
                data,
                Enum.Operation.Call
            ),
            "Could not execute token transfer"
        );
    }

    function getDomainSeparator() private view returns (bytes32) {

        uint256 chainId;
        assembly {
            chainId := chainid()
        }

        return keccak256(
            abi.encode(
                keccak256(
                    "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
                ),
                keccak256(bytes("TokenWithdrawModule")),
                keccak256(bytes("1")),
                chainId,
                address(this)
            )
        );
    }
}
```

### Testing the contract

#### Step 1: Start with an empty test file

Create a new file named `TokenTransferModule.test.ts` and include the following basic structure:

```typescript

> // Import necessary libraries and types

Reviewer 2: add all necessary imports here

describe("TokenTransferModule Tests", function () {

  // Define variables
   
  // Before hook to setup the contracts
  before(async () => {
  });

  // Add your test cases here
 it("Should successfully transfer tokens to bob", async function () {

 });
```

#### Step 2: Create test/utils/utils.ts file

Create a new file named `utils.ts` in the `test/utils` directory and include the code below. This file contains utility functions to generate the EIP-712 digest and execute transaction through the Safe account.

> ```typescript

Reviewer 2: This file is wordy :D Does it have to be so wordy? Do the digest and the other hashes need to be generated on the spot, or are they constant? Is there any way to shorten the code and make it more easy to understand?

Author: `PERMIT_TYPEHASH` can be read from module contract.
 Also, `DOMAIN_SEPARATOR`  can be read from module provide `getDomainSeparator` function is made external/public which is not a requirement for the contract to work.

Author: Removed `PERMIT_TYPEHASH`, `getDomainSeparator` and even `getDigest`.
Utils now only has `execTransaction` function.

    to,
    value,
    data,
    operation,
    0,
    0,
    0,
    ZeroAddress,
    ZeroAddress,
    nonce
  );

  let signatureBytes = "0x";
  const bytesDataHash = ethers.getBytes(transactionHash);

  // Get the addresses of the signers
  const addresses = await Promise.all(wallets.map(wallet => wallet.getAddress()));
  // Sort the signers by their addresses
  const sorted = wallets.sort((a, b) => {
    const addressA = addresses[wallets.indexOf(a)];
    const addressB = addresses[wallets.indexOf(b)];
    return addressA.localeCompare(addressB, "en", { sensitivity: "base" });
  });

  // Sign the transaction hash with each signer
  for (let i = 0; i < sorted.length; i++) {
    const flatSig = (await sorted[i].signMessage(bytesDataHash))
      .replace(/1b$/, "1f")
      .replace(/1c$/, "20");
    signatureBytes += flatSig.slice(2);
  }

  // Execute the transaction on the Safe contract
  await safe.execTransaction(
    to,
    value,
    data,
    operation,
    0,
    0,
    0,
    ZeroAddress,
    ZeroAddress,
    signatureBytes
  );
};

export {
  PERMIT_TYPEHASH,
  execTransaction,
  getDigest,
  getDomainSeparator,
};
```

> #### Step 3: Setup contracts and variables

Reviewer 2: Change the order, so that it is code first, and then explanation to be consistent.
Also, only include the before section, because the import scripts are added in the skeleton file.

1. **Import necessary libraries and types**:
   - Import the required libraries and types from `hardhat`, `chai`, and `ethers`.

2. **Define variables**:
   - Define variables to hold the signers, contract instances, and other necessary data.

3. **Setup signers and deploy contracts before running tests**:
   - Use the `before` hook to set up the signers and deploy the `Safe` and `SafeProxyFactory` contracts before running the tests.

4. **Helper function to setup contracts**:
   - Define a helper function `setupContracts` to deploy and initialize the contracts. This function:
     - Deploys the Safe contract and sets up the owners and threshold.
     - Deploys the TokenWithdrawModule contract.
     - Mints tokens to the Safe address.
     - Enables the TokenWithdrawModule in the Safe.
     - Verifies that the module is enabled in the Safe.

This setup ensures that each test starts with a fresh instance of the contracts and the necessary initial state.

```typescript
// Import necessary libraries and types
import { ethers } from "hardhat";
import { expect } from "chai";
import { Signer, ZeroAddress } from "ethers";

> import { Safe__factory, TestToken, TokenWithdrawModule } from "../typechain-types";

Reviewer 2: I get an error here. `Module '"../typechain-types"' has no exported member 'TestToken'.`

Author: resolved. Added `TestToken.sol` in tutorial

> import { SafeTransaction } from "@safe-global/safe-contracts";

Reviewer 2: Remove this line. This import is not used and the line is removed from the complete file.

    // Create the proxy with nonce
    await proxyFactory.createProxyWithNonce(
      await masterCopy.getAddress(),
      safeData,
      0n
    );

    if (safeAddress === ZeroAddress) {
      throw new Error("Safe address not found");
    }

    // Deploy the TokenWithdrawModule contract
    const exampleModule = await (
      await ethers.getContractFactory("TokenWithdrawModule", deployer)
    ).deploy(token.target, safeAddress);

    // Mint tokens to the safe address
    await token
      .connect(deployer)
      .mint(safeAddress, BigInt(10) ** BigInt(18) * BigInt(100000));

    const safe = await ethers.getContractAt("Safe", safeAddress);

    // Enable the module in the safe
    const enableModuleData = masterCopy.interface.encodeFunctionData(
      "enableModule",
      [exampleModule.target]
    );

    // Execute the transaction to enable the module
    await execTransaction(
      walletOwners.slice(0, threshold),
      safe,
      safe.target,
      0,
      enableModuleData,
      0
    );

    // Verify that the module is enabled
    expect(await safe.isModuleEnabled.staticCall(exampleModule.target)).to.be
      .true;

    return { exampleModule };
  };
});

```

> #### Step 4: Add test case

Reviewer 2: As above, change the order, so it is code first, and then the explanation.

    // Generate the digest for the transaction
    const digest = getDigest(
      "TokenWithdrawModule",
      await exampleModule.getAddress(),
      chainId,
      amount,
      await bob.getAddress(),
      nonce,
      deadline
    );

    const bytesDataHash = ethers.getBytes(digest);

    // Sign the digest with each wallet owner
    for (let i = 0; i < wallets.length; i++) {
      const flatSig = (await wallets[i].signMessage(bytesDataHash))
        .replace(/1b$/, "1f")
        .replace(/1c$/, "20");
      signatureBytes += flatSig.slice(2);
    }

    // Attempt to transfer tokens with an invalid signer (should fail)
    await expect(
      exampleModule
        .connect(charlie)
        .tokenTransfer(amount, await charlie.getAddress(), deadline, signatureBytes)
    ).to.be.revertedWith("GS026");

    // Transfer tokens with a valid signer
    await exampleModule
      .connect(bob)
      .tokenTransfer(amount, await bob.getAddress(), deadline, signatureBytes);

    // Verify the token balance of bob
    const balanceBob = await token.balanceOf.staticCall(await bob.getAddress());
    expect(balanceBob).to.be.equal(amount);
  });

});
```

### Run the tests

```bash

> npx harhdat test

Reviewer: ```suggestion
npx hardhat test
```
